File: Intersection of Two Linked Lists/Intersection_of_Two_Linked_Lists.py
# Runtime: 188 ms, faster than 99.67% of Python online submissions for Intersection of Two Linked Lists.
# Memory Usage: 40.9 MB, less than 100.00% of Python online submissions for Intersection of Two Linked Lists.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        lena, lenb = 0, 0

        temp = headA
        while temp:
            lena += 1
            temp = temp.next
        
        temp = headB
        while temp:
            lenb += 1
            temp = temp.next
        
        if lena is 0 or lenb is 0:
            return None

        if lena >= lenb:
            longer = headA
            shorter = headB
        else:
            longer = headB
            shorter = headA

        for i in range(abs(lena - lenb)):
            longer = longer.next
        
        while longer != shorter:
            longer = longer.next
            shorter = shorter.next

        return longer

File: Intersection of Two Linked Lists/Intersection_of_Two_Linked_Lists.cpp
// 思路一 记录第一个链表的长度和第二个链表的长度
// 然后让长链表的指针先移动几个，然后两个指针同时出发，这样相遇的时候就是交点

// Runtime: 56 ms, faster than 98.55% of C++ online submissions for Intersection of Two Linked Lists.
// Memory Usage: 16.8 MB, less than 100.00% of C++ online submissions for Intersection of Two Linked Lists.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) 
    {
        // 记录两个链表的长度
        int lena = 0, lenb = 0;
        for (ListNode *temp = headA; temp; temp = temp->next)
            lena++;
        for (ListNode *temp = headB; temp; temp = temp->next)    
            lenb++;
        if (lena == 0 || lenb == 0)
            return 0;
        
        ListNode *longer, *shorter;
        if (lena >= lenb)
            longer = headA, shorter = headB;
        else
            longer = headB, shorter = headA;
        
        for (int i = 0; i < abs(lena - lenb); i++)
            longer = longer->next;
        
        while (longer != shorter)
        {
            longer = longer->next;
            shorter = shorter->next;
        }
        return longer;
    }
};